Implement H5ESget requests function to retrieve requests from an event set

fortnern opened the pull request.

fortnern committed:
Implement H5ESget_requests() to retrieve request pointers and
corresponding connector IDs from an event set.  Add tests for this
function.

github-actions[bot] committed:
Committing clang-format changes

fortnern committed:
Add "order" parameter to H5ESget_requests, to allow the user to specify
the order in which requests are returned.

github-actions[bot] committed:
Committing clang-format changes

fortnern committed:
Fix bugs with H5ESinsert_request()

github-actions[bot] committed:
Committing clang-format changes

fortnern committed:
Add RELEASE.txt note for H5ESget_requests()

fortnern committed:
Merge branch 'h5esget_requests' of github.com:fortnern/hdf5 into h5esget_requests

epourmal reviewed (approved):
The new function needs Doxygen section. We will need to discuss how to present developers' functions in a separate section of RM.

soumagne commented on src/H5ES.c:
```diff
@@ -235,6 +235,61 @@ H5ESget_op_counter(hid_t es_id, uint64_t *op_counter /*out*/)
+    /* Retrieve the count, if non-NULL */
+    if (count)
+        *count = H5ES__list_count(&es->active);
+
```
Not sure if you should allow for count to be NULL if the routine does nothing if count is NULL :)

fortnern replied:
Well free() will happily do nothing if you pass NULL ;)  Though I don't have a strong feeling one way or the other.

In general though, I should mention that I thought this coding pattern (passing a pointer to the length of a buffer that is updated during the routine) was common throughout the library, but now I can't find any examples of it in the public API. Perhaps it is only used for internal interfaces. I'm open to changing the interface, though having the function return both the data and the length of the data in a single call may prevent some race conditions in a multithreaded application (with threadsafe on). Certainly we should always pass the length of the user's buffer.

fortnern replied:
API revised according to discussion on Friday. There are now separate input and output parameters array_len and count. The function now does something even if array_len is 0 (it returns count). I don't know if it's worth throwing an error if array_len is 0 and count is NULL.

fortnern commented:
> 

Is the doxygen section the comments stuff I see in the public headers above the prototypes? There isn't one for the other function in H5ESdevelop.h, should I add that? Some of the other develop headers are also missing them. Are developer functions supposed to have doxygen entries?

fortnern committed:
Change H5ESget_requests() API to have separate input and output
parameters for the allocated array length(s) (in) and the number of events in
the event set (out).

github-actions[bot] committed:
Committing clang-format changes

fortnern commented:
Github is apparently broken. The code checks passed, and the format check should pass as well since the auto check-in, but github won't run it. Let me know if it looks good and Larry can merge it manually.

soumagne commented on src/H5ES.c:
```diff
@@ -235,6 +235,61 @@ H5ESget_op_counter(hid_t es_id, uint64_t *op_counter /*out*/)
+ */
+herr_t
+H5ESget_requests(hid_t es_id, H5_iter_order_t order, hid_t *connector_ids, void **requests, size_t array_len,
+                 size_t *count /*out*/)
```
Looks fine, I think otherwise it's usually better to just name it `max_count`, similar to what `epoll_wait()` does: https://man7.org/linux/man-pages/man2/epoll_wait.2.html
That just makes it clearer that *count <= array_len

soumagne replied:
OK forget my last comment, reading more I see it's kind of decoupled anyway and *count can be greater than array_len.
